wavesrv/pkg/cmdrunner/cmdrunner.go

```diff
@@ -4776,6 +4818,154 @@ func MarkdownViewCommand(ctx context.Context, pk *scpacket.FeCommandPacketType)
 	return update, nil
 }
 
+func StatDir(ctx context.Context, ids resolvedIds, path string, fileCallback func(pk *packet.FileStatPacketType, done bool, err error)) {
+	log.Printf("running %v: \n", path)
+	statPk, _, err := getFileStat(ctx, ids, path)
+	if err != nil {
+		fileCallback(nil, true, err)
+		return
+	}
+	if !statPk.IsDir {
+		fileCallback(statPk, true, nil)
+		return
+	} else {
+		cwd := ids.Remote.FeState["cwd"]
+		listDirPk := packet.MakeListDirPacket()
+		listDirPk.ReqId = uuid.New().String()
+		if filepath.IsAbs(path) {
+			listDirPk.Path = path
+		} else {
+			listDirPk.Path = filepath.Join(cwd, path)
+		}
+		msh := ids.Remote.MShell
+		listDirIter, err := msh.ListDir(ctx, listDirPk)
+		if err != nil {
+			fileCallback(nil, true, err)
+			return
+		}
+		defer listDirIter.Close()
+		for {
+			respIf, err := listDirIter.Next(ctx)
+			if err != nil {
+				fileCallback(nil, true, err)
+				return
+			}
+			resp, ok := respIf.(*packet.FileStatPacketType)
+			if !ok || resp == nil {
+				fileCallback(nil, true, fmt.Errorf("error unmarshalling file stat response type %v %v", resp, respIf))
+				return
+			}
+			if resp.Error != "" {
+				err = fmt.Errorf(resp.Error)
+			} else {
+				err = nil
+			}
+			fileCallback(resp, resp.Done, err)
+			if resp.Done {
+				return
+			}
+		}
+	}
+}
+
+func getCurCirFileOffset(ctx context.Context, cmd *sstore.CmdType) (int64, error) {
+	ptyOutFileName, err := scbase.PtyOutFile(cmd.ScreenId, cmd.LineId)
+	if err != nil {
+		return 0, err
+	}
+	stat, err := cirfile.StatCirFile(ctx, ptyOutFileName)
+	if err != nil {
+		return 0, err
+	}
+	return stat.FileOffset, nil
+}
+
+func ViewDirCommand(ctx context.Context, pk *scpacket.FeCommandPacketType) (sstore.UpdatePacket, error) {
+	ids, err := resolveUiIds(ctx, pk, R_Session|R_Screen|R_RemoteConnected)
+	if err != nil {
+		return nil, err
+	}
+	path := ids.Remote.FeState["cwd"]
+	if len(pk.Args) > 0 {
+		path = pk.Args[0]
+	}
+	cmd, err := sstore.GetCmdByScreenId(ctx, pk.Kwargs["screenid"], pk.Kwargs["lineid"])
```

Comment:
currently this doesn't conform with the "outputpty" paradigm. Need to implement a version of this that outputs to a line if "outputpty" is not specified, because I think this might be useful to a user as well